karapace/kafka_rest_apis/consumer_manager.py

```diff
@@ -495,47 +505,44 @@ async def fetch(self, internal_name: Tuple[str, str], content_type: str, formats
+                read_buffered = True
             LOG.info(
                 "Gathered %d total messages (%d bytes read) in %r",
                 message_count,
                 read_bytes,
                 time.monotonic() - start_time,
             )
-            for tp in poll_data:
-                for msg in poll_data[tp]:
-                    try:
-                        key = await self.deserialize(msg.key, request_format) if msg.key else None
-                    except DeserializationError as e:
-                        KarapaceBase.unprocessable_entity(
-                            message=f"key deserialization error for format {request_format}: {e}",
-                            sub_code=RESTErrorCodes.HTTP_UNPROCESSABLE_ENTITY.value,
-                            content_type=content_type,
-                        )
-                    try:
-                        value = await self.deserialize(msg.value, request_format) if msg.value else None
-                    except DeserializationError as e:
-                        KarapaceBase.unprocessable_entity(
-                            message=f"value deserialization error for format {request_format}: {e}",
-                            sub_code=RESTErrorCodes.HTTP_UNPROCESSABLE_ENTITY.value,
-                            content_type=content_type,
-                        )
-                    element = {
-                        "topic": tp.topic,
-                        "partition": tp.partition,
-                        "offset": msg.offset,
-                        "timestamp": msg.timestamp,
-                        "key": key,
-                        "value": value,
-                    }
-                    response.append(element)
+            for msg in poll_data:
+                try:
+                    key = await self.deserialize(msg.key(), request_format) if msg.key() else None
+                except DeserializationError as e:
+                    KarapaceBase.unprocessable_entity(
+                        message=f"key deserialization error for format {request_format}: {e}",
+                        sub_code=RESTErrorCodes.HTTP_UNPROCESSABLE_ENTITY.value,
+                        content_type=content_type,
+                    )
+                try:
+                    value = await self.deserialize(msg.value(), request_format) if msg.value() else None
+                except DeserializationError as e:
+                    KarapaceBase.unprocessable_entity(
+                        message=f"value deserialization error for format {request_format}: {e}",
+                        sub_code=RESTErrorCodes.HTTP_UNPROCESSABLE_ENTITY.value,
+                        content_type=content_type,
+                    )
+                element = {
+                    "topic": msg.topic(),
+                    "partition": msg.partition(),
+                    "offset": msg.offset(),
+                    "timestamp": msg.timestamp()[1],
```

Comment:
mmm the `[1]` is non trivial, could you add a comment like:
`msg.timestamp()[1] # returns a tuple of message timestamp type, and timestamp we are interested only on the latter.`

Also we don't need to parse the type of timing and do different translation based on the type of time?
Should we maybe hide this detail behind a logic and avoid to forwarding that weird type outside of your wrapper?

Reply:
I would also explictly check that the `[0]` element its the value expected :)